File: app/src/main/java/io/eodc/planit/fragment/ModifyAssignmentFragment.java
package io.eodc.planit.fragment;

import android.annotation.SuppressLint;
import android.app.DatePickerDialog;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.Spinner;

import com.google.common.collect.Iterables;

import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.eodc.planit.R;
import io.eodc.planit.adapter.AssignmentType;
import io.eodc.planit.adapter.AssignmentTypeAdapter;
import io.eodc.planit.db.Assignment;
import io.eodc.planit.db.Class;
import io.eodc.planit.db.PlannerDatabase;
import io.eodc.planit.helper.AssignmentInfoInputHelper;
import io.eodc.planit.helper.KeyboardFocusManager;
import io.eodc.planit.model.ClassListViewModel;

/**
 * Created by 2n on 5/16/18.
 */

public class ModifyAssignmentFragment extends DialogFragment implements
        DatePickerDialog.OnDateSetListener,
        AdapterView.OnItemSelectedListener {

    @BindView(R.id.btn_restore)                 Button              mBtnRestore;
    @BindView(R.id.spinner_type)                Spinner             mSpinnerType;
    @BindView(R.id.spinner_class)               Spinner             mSpinnerClass;
    @BindView(R.id.edit_title)        EditText            mEditTitle;
    @BindView(R.id.edit_due_date)               EditText            mEditDueDate;
    @BindView(R.id.edit_notes)                  EditText            mEditNotes;
    @BindView(R.id.layout_edit_title) TextInputLayout     mLayoutEditTitle;
    @BindView(R.id.layout_edit_due)        TextInputLayout     mLayoutEditDueDate;

    private Assignment  mAssignment;

    private long    mSelectedClassId = 1;
    private int     mSelectedType;
    private int     mDueDay;
    private int     mDueMonth;
    private int     mDueYear;

    public static ModifyAssignmentFragment newInstance(Assignment assignment) {
        ModifyAssignmentFragment fragment = new ModifyAssignmentFragment();
        fragment.mAssignment = assignment;
        return fragment;
    }

    @OnClick(R.id.btn_cancel)
    void dismissDialog() {
        dismiss();
    }

    @SuppressLint("StaticFieldLeak")
    @OnClick(R.id.btn_confirm)
    void editAssignment() {
        String title = mEditTitle.getText().toString().trim();
        String notes = mEditNotes.getText().toString().trim();
        String dueDate = mEditDueDate.getText().toString().trim();

        if (title.equals("") || mEditDueDate.getText().toString().trim().equals("")) {
            if (title.equals("")) {
                mLayoutEditTitle.setError("Title cannot be empty.");
            }
            if (dueDate.equals("")) {
                mLayoutEditDueDate.setError("Due date can't be empty");
            }
        } else {
            mAssignment.setTitle(title);
            mAssignment.setNotes(notes);
            mAssignment.setClassId((int) mSelectedClassId);
            switch (mSelectedType) {
                case 0:
                    mAssignment.setType(Assignment.TYPE_HOMEWORK);
                    break;
                case 1:
                    mAssignment.setType(Assignment.TYPE_TEST);
                    break;
                case 2:
                    mAssignment.setType(Assignment.TYPE_PROJECT);
                    break;
            }

            mAssignment.setDueDate(new DateTime(mDueYear, mDueMonth, mDueDay, 0, 0));

            new Thread(() -> PlannerDatabase.getInstance(getContext()).assignmentDao().updateAssignment(mAssignment)).start();
            dismiss();
        }
    }

    @OnClick(R.id.edit_due_date)
    void showDatePicker() {
        DatePickerFragment fragment = DatePickerFragment.newInstance(this, mDueYear, mDueMonth, mDueDay);

        if (getFragmentManager() != null) fragment.show(getFragmentManager(), null);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getDialog() != null && getDialog().getWindow() != null) {
            int width = (int) getResources().getDimension(R.dimen.edit_dialog_width);
            int height = (int) getResources().getDimension(R.dimen.edit_dialog_height);
            getDialog().getWindow().setLayout(width, height);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.dialog_edit_assignment, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ViewModelProviders.of(this).get(ClassListViewModel.class).getClasses()
                .observe(this, this::setupClassSpinner);

    }

    private void setupInputListeners() {
        mEditTitle.addTextChangedListener(new AssignmentInfoInputHelper(mLayoutEditTitle));
        if (getView() != null) {
            getView().setOnTouchListener((v, motionEvent) -> {
                switch (motionEvent.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        KeyboardFocusManager.clearTextFocus(getContext(), v,
                                mEditDueDate, mEditNotes, mEditTitle);
                        break;
                    case MotionEvent.ACTION_UP:
                        v.performClick();
                        break;
                }
                return true;
            });
        }
    }

    private void setupTypeSpinner() {
        if (getContext() != null) {
            List<AssignmentType> types = new ArrayList<AssignmentType>() {{
                add(new AssignmentType("Homework", R.drawable.ic_homework_black_24dp));
                add(new AssignmentType("Quiz/Test", R.drawable.ic_test_black_24dp));
                add(new AssignmentType("Project", R.drawable.ic_group_black_24dp));
            }};

            AssignmentTypeAdapter typeAdapter = new AssignmentTypeAdapter(getContext(),
                    R.layout.item_assignment_type,
                    R.id.text_title,
                    types);
            mSpinnerType.setAdapter(typeAdapter);
            mSpinnerType.setOnItemSelectedListener(this);
        }
    }

    private void setupClassSpinner(List<Class> classes) {
        if (getContext() != null) {
            ArrayAdapter mClassAdapter = new ArrayAdapter<>(getContext(),
                    android.R.layout.simple_spinner_dropdown_item,
                    android.R.id.text1,
                    classes);
            mSpinnerClass.setAdapter(mClassAdapter);
            mSpinnerClass.setOnItemSelectedListener(this);

            mSpinnerClass.setSelection(Iterables.indexOf(classes,
                    c -> c.getId() == mAssignment.getClassId()));

            setupTypeSpinner();
            fillAssignmentInfo();
            setupInputListeners();
        }
    }

    private void fillAssignmentInfo() {
        DateTime dtDue = mAssignment.getDueDate();
        mEditTitle.setText(mAssignment.getTitle());
        mEditDueDate.setText(dtDue.toString("dd MMM, YYYY"));
        mEditNotes.setText(mAssignment.getNotes());

        mDueDay = dtDue.getDayOfMonth();
        mDueMonth = dtDue.getMonthOfYear();
        mDueYear = dtDue.getYear();
    }

    @Override
    public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
        mDueDay = dayOfMonth;
        mDueMonth = month + 1;
        mDueYear = year;

        DateTime dueDate = new DateTime(mDueYear, mDueMonth, mDueDay, 0, 0);

        mEditDueDate.setText(dueDate.toString(getString(R.string.due_date_pattern)));
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        if (parent.equals(mSpinnerType)) mSelectedType = position;
        else if (parent.equals(mSpinnerClass)) {
            Class selected = (Class) parent.getSelectedItem();
            mSelectedClassId = selected.getId();
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) { }
}

File: app/src/main/java/io/eodc/planit/fragment/ModifyAssignmentFragment.kt
package io.eodc.planit.fragment

import android.app.DatePickerDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.DatePicker
import com.google.common.collect.Iterables
import io.eodc.planit.R
import io.eodc.planit.activity.MainActivity
import io.eodc.planit.adapter.AssignmentType
import io.eodc.planit.adapter.AssignmentTypeAdapter
import io.eodc.planit.db.Assignment
import io.eodc.planit.db.PlannerDatabase
import io.eodc.planit.db.Subject
import io.eodc.planit.helper.AssignmentInfoInputHelper
import io.eodc.planit.helper.KeyboardFocusManager
import kotlinx.android.synthetic.main.dialog_edit_assignment.*
import org.joda.time.DateTime

/**
 * Created by 2n on 5/16/18.
 */

class ModifyAssignmentFragment : androidx.fragment.app.DialogFragment(), DatePickerDialog.OnDateSetListener, AdapterView.OnItemSelectedListener {
    private lateinit var mAssignment: Assignment

    private var mSelectedClassId: Long = 1
    private var mSelectedType: Int = 0
    private var mDueDay: Int = 0
    private var mDueMonth: Int = 0
    private var mDueYear: Int = 0

    private fun editAssignment() {
        val title = editTitle.text.toString().trim { it <= ' ' }
        val notes = editNotes.text.toString().trim { it <= ' ' }
        val dueDate = editDue.text.toString().trim { it <= ' ' }

        if (title == "" || editDue.text.toString().trim { it <= ' ' } == "") {
            if (title == "") {
                layoutEditTitle.error = "Title cannot be empty."
            }
            if (dueDate == "") {
                layoutEditDue!!.error = "Due date can't be empty"
            }
        } else {
            mAssignment.title = title
            mAssignment.notes = notes
            mAssignment.classId = mSelectedClassId.toInt()
            when (mSelectedType) {
                0 -> mAssignment.type = Assignment.TYPE_HOMEWORK
                1 -> mAssignment.type = Assignment.TYPE_TEST
                2 -> mAssignment.type = Assignment.TYPE_PROJECT
            }

            mAssignment.dueDate = DateTime(mDueYear, mDueMonth, mDueDay, 0, 0)

            Thread { context?.let { PlannerDatabase.getInstance(it) }
                    ?.assignmentDao()
                    ?.updateAssignment(mAssignment) }.start()
            dismiss()
        }
    }

    private fun showDatePicker() {
        val fragment = DatePickerFragment.newInstance(this, mDueYear, mDueMonth, mDueDay)

        if (fragmentManager != null) fragment.show(fragmentManager!!, null)
    }

    override fun onResume() {
        super.onResume()
        if (dialog != null && dialog.window != null) {
            val width = resources.getDimension(R.dimen.edit_dialog_width).toInt()
            val height = resources.getDimension(R.dimen.edit_dialog_height).toInt()
            dialog.window!!.setLayout(width, height)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.dialog_edit_assignment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupClassSpinner()
    }

    private fun setupInputListeners() {
        editTitle.addTextChangedListener(AssignmentInfoInputHelper(layoutEditTitle))
        if (view != null) {
            view!!.setOnTouchListener { v, motionEvent ->
                when (motionEvent.action) {
                    MotionEvent.ACTION_DOWN -> KeyboardFocusManager.clearTextFocus(context, v,
                            editDue, editNotes, editTitle)
                    MotionEvent.ACTION_UP -> v.performClick()
                }
                true
            }
        }

        btnCancel.setOnClickListener { it ->
            Thread {
                context?.let { PlannerDatabase.getInstance(it) }
                        ?.assignmentDao()
                        ?.updateAssignment(mAssignment)
            }.start()
            dismiss()
        } // Force a redraw of the recyclerview
        btnConfirm.setOnClickListener { editAssignment() }
        editDue.setOnClickListener { showDatePicker() }
    }

    private fun setupTypeSpinner() {
        if (context != null) {
            val types = arrayListOf(
                    AssignmentType("Homework", R.drawable.ic_homework_black_24dp),
                    AssignmentType("Quiz/Test", R.drawable.ic_test_black_24dp),
                    AssignmentType("Project", R.drawable.ic_group_black_24dp))

            val typeAdapter = AssignmentTypeAdapter(context!!,
                    R.layout.item_assignment_type,
                    R.id.textHeaderTitle,
                    types)
            spinnerType.adapter = typeAdapter
            spinnerType.onItemSelectedListener = this

            val position =
                    when (mAssignment.type) {
                        Assignment.TYPE_HOMEWORK -> 0
                        Assignment.TYPE_TEST -> 1
                        Assignment.TYPE_PROJECT -> 2
                        else -> 0
                    }
            spinnerType.setSelection(position)
        }
    }

    private fun setupClassSpinner() {
        if (context != null && activity != null) {
            val subjects = (activity as MainActivity).subjects
            val mClassAdapter = ArrayAdapter(context!!,
                    android.R.layout.simple_spinner_dropdown_item,
                    android.R.id.text1,
                    subjects!!)
            spinnerSubject.adapter = mClassAdapter
            spinnerSubject.onItemSelectedListener = this

            spinnerSubject.setSelection(Iterables.indexOf(subjects) { c ->
                c?.id == mAssignment.classId
            })

            setupTypeSpinner()
            fillAssignmentInfo()
            setupInputListeners()
        }
    }

    private fun fillAssignmentInfo() {
        val dtDue = mAssignment.dueDate
        editTitle.setText(mAssignment.title)
        editDue.setText(dtDue.toString("dd MMM, YYYY"))
        editNotes.setText(mAssignment.notes)

        mDueDay = dtDue.dayOfMonth
        mDueMonth = dtDue.monthOfYear
        mDueYear = dtDue.year
    }

    override fun onDateSet(view: DatePicker, year: Int, month: Int, dayOfMonth: Int) {
        mDueDay = dayOfMonth
        mDueMonth = month + 1
        mDueYear = year

        val dueDate = DateTime(mDueYear, mDueMonth, mDueDay, 0, 0)

        editDue.setText(dueDate.toString(getString(R.string.due_date_pattern)))
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        if (parent == spinnerType)
            mSelectedType = position
        else if (parent == spinnerSubject) {
            val selected = parent.selectedItem as Subject
            mSelectedClassId = selected.id.toLong()
        }
    }

    override fun onNothingSelected(parent: AdapterView<*>) {}

    companion object {

        fun newInstance(assignment: Assignment): ModifyAssignmentFragment {
            val fragment = ModifyAssignmentFragment()
            fragment.mAssignment = assignment
            return fragment
        }
    }
}
